bin/reth/src/builder/mod.rs

```diff
@@ -1356,28 +1360,47 @@ pub struct NodeHandle {
 
     /// The receiver half of the channel for the consensus engine.
     /// This can be used to wait for the consensus engine to exit.
-    consensus_engine_rx: oneshot::Receiver<Result<(), BeaconConsensusEngineError>>,
+    pub consensus_engine_rx: oneshot::Receiver<Result<(), BeaconConsensusEngineError>>,
 
     /// Flag indicating whether the node should be terminated after the pipeline sync.
-    terminate: bool,
+    pub terminate: bool,
 }
 
 impl NodeHandle {
     /// Returns the [RethRpcServerHandles] for this node.
     pub fn rpc_server_handles(&self) -> &RethRpcServerHandles {
         &self.rpc_server_handles
     }
+}
 
-    /// Waits for the node to exit, if it was configured to exit.
-    pub async fn wait_for_node_exit(self) -> eyre::Result<()> {
-        self.consensus_engine_rx.await??;
+/// A Future which resolves when the node exits
+#[derive(Debug)]
+pub struct NodeExitFuture {
+    /// The receiver half of the channel for the consensus engine.
+    /// This can be used to wait for the consensus engine to exit.
+    pub consensus_engine_rx: oneshot::Receiver<Result<(), BeaconConsensusEngineError>>,
 
-        if self.terminate {
-            Ok(())
-        } else {
-            // The pipeline has finished downloading blocks up to `--debug.tip` or
-            // `--debug.max-block`. Keep other node components alive for further usage.
-            futures::future::pending().await
+    /// Flag indicating whether the node should be terminated after the pipeline sync.
+    pub terminate: bool,
+}
+
+impl Future for NodeExitFuture {
+    type Output = eyre::Result<()>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let pin = self.get_mut();
+
+        match pin.consensus_engine_rx.poll_unpin(cx) {
```

Comment:
this needs to handle the case where the oneshot already returned Ready, and should be be polled again, relevant, see poll impl of https://docs.rs/futures/latest/futures/future/struct.Fuse.html